default.nix

```diff
@@ -212,6 +212,20 @@ let
       # require n-i >= 0.4. dontCheck helps us avoid conflicts with
       # neat-interpolation's test dependencies.
       neat-interpolation = pkgs.haskell.lib.dontCheck super.neat-interpolation_0_5_1;
+
+      # 2020-07-23 hnix uses multiple functions that are unavailable in
+      # data-fix < 0.3.
+      data-fix = haskellPackages.callPackage
+       ({ mkDerivation, base, deepseq, hashable, stdenv }:
+        mkDerivation {
+          pname = "data-fix";
+          version = "0.3.0";
+          sha256 = "9e59b3ed694b5139316093b3767842e60ad4821858459e7cd763e5773dfa99a0";
+          libraryHaskellDepends = [ base deepseq hashable ];
+          homepage = "https://github.com/spell-music/data-fix";
+          description = "Fixpoint data types";
+          license = stdenv.lib.licenses.bsd3;
+        }) {};
```

Comment:
I arrived at this setup after a bit of trial and error. If there's a better way to achieve the same, I'd happily use it.

Reply:
Maybe try `haskellPackages.callHackage`:

Given example is:
```text
...
foo = self.callHackage "foo" "1.5.3" {};
...
```

It is much more clear and obvious what it does.

It calls official `hackage2nix` machinery into motion and takes Haskell package and automatically generates Nix Haskell package.

`callHackage` is wonderful to use in cases when Nixpkgs dodges to follow the upstream Hackage.

Reply:
I'm getting an error with this patch:

```diff
--- a/default.nix
+++ b/default.nix
@@ -215,17 +215,7 @@ let
 
       # 2020-07-23 hnix uses multiple functions that are unavailable in
       # data-fix < 0.3.
-      data-fix = haskellPackages.callPackage
-       ({ mkDerivation, base, deepseq, hashable, stdenv }:
-        mkDerivation {
-          pname = "data-fix";
-          version = "0.3.0";
-          sha256 = "9e59b3ed694b5139316093b3767842e60ad4821858459e7cd763e5773dfa99a0";
-          libraryHaskellDepends = [ base deepseq hashable ];
-          homepage = "https://github.com/spell-music/data-fix";
-          description = "Fixpoint data types";
-          license = stdenv.lib.licenses.bsd3;
-        }) {};
+      data-fix = haskellPackages.callHackage "data-fix" "0.3.0" {};
     };
```

```
$ nix-build --argstr compiler ghc883 --arg buildFromSdist true
building '/nix/store/h8zl5khhsf9x2w86qgqw1f48g2y8fxhv-cabal2nix-hnix.drv'...
installing
building '/nix/store/zknicl0518i1m0sh7f7yfk0dby6dsv3q-all-cabal-hashes-component-data-fix-0.3.0.drv'...
tar: */data-fix/0.3.0/data-fix.json: Not found in archive
tar: */data-fix/0.3.0/data-fix.cabal: Not found in archive
tar: Exiting with failure status due to previous errors
builder for '/nix/store/zknicl0518i1m0sh7f7yfk0dby6dsv3q-all-cabal-hashes-component-data-fix-0.3.0.drv' failed with exit code 2
cannot build derivation '/nix/store/9h8v1aia6ak6fb29ljc356aissz0rzbz-cabal2nix-data-fix-0.3.0.drv': 1 dependencies couldn't be built
error: build of '/nix/store/9h8v1aia6ak6fb29ljc356aissz0rzbz-cabal2nix-data-fix-0.3.0.drv' failed
(use '--show-trace' to show detailed location information)
```

Any ideas?

Reply:
AFAIK, it means that Nixpkgs does not imported/updated Hackage revision information into the system.

Seems like `callHackageDirect` should be used.

I am looking into it, you also can look into `haskell-modules/make-package-set.nix` to understand this.

Reply:
> Seems like `callHackageDirect` should be used.

Do you mean this?

```diff
--- a/default.nix
+++ b/default.nix
@@ -215,17 +215,11 @@ let
 
       # 2020-07-23 hnix uses multiple functions that are unavailable in
       # data-fix < 0.3.
-      data-fix = haskellPackages.callPackage
-       ({ mkDerivation, base, deepseq, hashable, stdenv }:
-        mkDerivation {
-          pname = "data-fix";
-          version = "0.3.0";
-          sha256 = "9e59b3ed694b5139316093b3767842e60ad4821858459e7cd763e5773dfa99a0";
-          libraryHaskellDepends = [ base deepseq hashable ];
-          homepage = "https://github.com/spell-music/data-fix";
-          description = "Fixpoint data types";
-          license = stdenv.lib.licenses.bsd3;
-        }) {};
+      data-fix = haskellPackages.callHackageDirect {
+        pkg = "data-fix";
+        ver = "0.3.0";
+        sha256 = "9e59b3ed694b5139316093b3767842e60ad4821858459e7cd763e5773dfa99a0";
+      } {};
     };
 
     modifier = drv: pkgs.haskell.lib.overrideCabal drv (attrs: {
```

That produces a different error:

```
$ nix-build --argstr compiler ghc883 --arg buildFromSdist true
building '/nix/store/dkf1vxj712fmq58wjl03vw2px39n36dk-cabal2nix-hnix.drv'...
installing
building '/nix/store/widy86ykjxr8mviq5c9kkj4alvcgnq4i-source.drv'...

trying http://hackage.haskell.org/package/data-fix-0.3.0/data-fix-0.3.0.tar.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  5516  100  5516    0     0  61977      0 --:--:-- --:--:-- --:--:-- 61288
unpacking source archive /tmp/nix-build-source.drv-0/data-fix-0.3.0.tar.gz
hash mismatch in fixed-output derivation '/nix/store/cg310v82g2pnmi6gyp2dhd544jdnxhcj-source':
  wanted: sha256:184rz8ypgrb3sxy9wiaq321d82p689w7dcwkc0qkjlabd7nv6ncy
  got:    sha256:1w3z4fa28zxqwgsynlz025rkmrdcv7bj66031l502nc3x3yfljn6
cannot build derivation '/nix/store/khmfa7xv97fbziibldgi120dccjfkxqj-cabal2nix-data-fix.drv': 1 dependencies couldn't be built
error: build of '/nix/store/khmfa7xv97fbziibldgi120dccjfkxqj-cabal2nix-data-fix.drv' failed
(use '--show-trace' to show detailed location information)
```

Reply:
```nix
      data-fix = haskellPackages.callHackageDirect {
        pkg = "dafa-fix";
        ver = "0.3.0";
        sha256 = "184rz8ypgrb3sxy9wiaq321d82p689w7dcwkc0qkjlabd7nv6ncy";
      };
```

Did not work, nor `self.callHackageDirect`.

At this point I think, lets merge it as you presented it working and move forward - it is a temporary change and we can and would rehash it in any case.